src/Storages/StorageS3.cpp

```diff
@@ -327,6 +345,167 @@ class StorageS3Sink : public SinkToStorage
+            auto value = key_column->getDataAt(row);
+            validatePartitionKey(value);
+            auto [it, inserted] = sub_chunks_indices.emplace(value, sub_chunks_indices.size());
+            selector.push_back(it->second);
+        }
+
+        Chunks sub_chunks;
+        sub_chunks.reserve(sub_chunks_indices.size());
+        for (size_t column_index = 0; column_index < columns.size(); ++column_index)
+        {
+            MutableColumns column_sub_chunks = columns[column_index]->scatter(sub_chunks_indices.size(), selector);
+            if (column_index == 0) /// Set sizes for sub-chunks.
+            {
+                for (size_t sub_chunk_index = 0; sub_chunk_index < column_sub_chunks.size(); ++sub_chunk_index)
+                {
+                    sub_chunks.emplace_back(Columns(), column_sub_chunks[sub_chunk_index]->size());
+                }
+            }
+            for (size_t sub_chunk_index = 0; sub_chunk_index < column_sub_chunks.size(); ++sub_chunk_index)
+            {
+                sub_chunks[sub_chunk_index].addColumn(std::move(column_sub_chunks[sub_chunk_index]));
+            }
+        }
+
+        for (const auto & [partition_id, sub_chunk_index] : sub_chunks_indices)
+        {
+            getSinkForPartition(partition_id)->consume(std::move(sub_chunks[sub_chunk_index]));
+        }
+    }
+
+    void onFinish() override
+    {
+        for (auto & [partition_id, sink] : sinks)
+        {
+            sink->onFinish();
+        }
+    }
+
+private:
+    using SinkPtr = std::shared_ptr<StorageS3Sink>;
+
+    const String format;
+    const Block sample_block;
+    ContextPtr context;
+    const CompressionMethod compression_method;
+    std::shared_ptr<Aws::S3::S3Client> client;
+    const String bucket;
+    const String key;
+    size_t min_upload_part_size;
+    size_t max_single_part_upload_size;
+
+    ExpressionActionsPtr partition_by_expr;
+    String partition_by_column_name;
+
+    std::unordered_map<String, SinkPtr> sinks;
+
+    static String replaceWildcards(const String & haystack, const String & partition_id)
+    {
+        return boost::replace_all_copy(haystack, PARTITION_ID_WILDCARD, partition_id);
+    }
+
+    SinkPtr getSinkForPartition(const String & partition_id)
+    {
+        if (sinks.count(partition_id) == 0)
+        {
+            sinks.emplace(partition_id, std::make_shared<StorageS3Sink>(
+                format,
+                sample_block,
+                context,
+                compression_method,
+                client,
+                replaceWildcards(bucket, partition_id),
+                replaceWildcards(key, partition_id),
+                min_upload_part_size,
+                max_single_part_upload_size
+            ));
+        }
+
+        return sinks[partition_id];
+    }
+
+    static void validatePartitionKey(const StringRef & str)
+    {
+        const char * end = str.data + str.size;
+
+        const char * first_symbol = find_first_symbols<
```

Comment:
But actually simple loop and checking `*pos < 32 && *pos != '{' && *pos != '}' && ...` or something like this will be better, because in current implementation we perform 3 iteration over whole string on "happy path"

Reply:
```
vladimir@cristal:~/find_first$ cat 1.cpp 
int main(int argc, char ** argv)
{
    bool result;
    for (int j = 0; j < 100000000; ++j)
    {
        result = true;
        for (char * i = argv[0]; *i; ++i)
        {
            if (*i < 32 || *i == '{' || *i == '}' || *i == '*' || *i == '?')
            {
                result = false;
                break;
            }
        }
    }
    return result ? 0 : 1;
}

```

```
#include "/home/vladimir/ClickHouse9/base/common/find_symbols.h"

int main(int argc, char ** argv)
{
    bool result;
    size_t len = 0;
    for (; argv[0][len] == 0; ++len);
    char *end = argv[0]+len;
    for (int j = 0; j < 100000000; ++j)
    {
        result = true;

        const char * first_symbol = find_first_symbols<
                '\x00', '\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\x07',
                '\x08', '\x09', '\x0A', '\x0B', '\x0C', '\x0D', '\x0E', '\x0F'>(argv[0], end);
        if (first_symbol == end)
        {
            first_symbol = find_first_symbols<
                    '\x10', '\x11', '\x12', '\x13', '\x14', '\x15', '\x16', '\x17',
                    '\x18', '\x19', '\x1A', '\x1B', '\x1C', '\x1D', '\x1E', '\x1F'
                >(argv[0], end);
            if (first_symbol == end)
            {
                first_symbol = find_first_symbols<'{', '}', '*', '?'>(argv[0], end);
                if (first_symbol == end)
                {
                    result = false;
                }
            }
        }
    }
    return result ? 0 : 1;
}
```

```
vladimir@cristal:~/find_first$ g++ -std=c++17 -O2 1.cpp -o 1
vladimir@cristal:~/find_first$ g++ -std=c++17 -O2 2.cpp -o 2
vladimir@cristal:~/find_first$ time ./1 abcdefghjkja

real	0m0.378s
user	0m0.367s
sys	0m0.000s
vladimir@cristal:~/find_first$ time ./2 abcdefghjkja

real	0m0.115s
user	0m0.114s
sys	0m0.000s
```

Reply:
Okay